nixos/qemu-vm: add option to specify security model to use for a shared directory

ajamtli opened the pull request:
## Description of changes

The security model for shared directories are hard coded to 'none'. This patch adds an option to specify the security model to use for the directory.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
      - qemu-vm-restrictnetwork: passed
      - qemu-vm-volatile-root: passed
      - qemu-vm-external-disk-image: passed
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)

      Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
      1 package blacklisted: nixos-install-tools

- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ajamtli committed:
nixos/qemu-vm: add option to specify security model to use for a shared directory

rnhmjoj commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -474,6 +474,11 @@ in
+            options.securityModel = mkOption {
+              type = types.enum [ "passthrough" "mapped" "mapped-xattr" "mapped-file" "none" ];
+              default = "none";
+              description = lib.mdDoc "The security model to use for this share";
```
```suggestion
              description = lib.mdDoc ''
                The security model to use for this share:

                - `passthrough`: files are stored using the same credentials as they are created on the guest (this requires QEMU to run as root)
                - `mapped-xattr`: some of the file attributes like uid, gid, mode bits and link target are stored as file attributes
                - `mapped-file`: the attributes are stored in the hidden .virtfs_metadata directory. Directories exported by this security model cannot interact with other unix tools
                - `none`: same as "passthrough" except the sever won't report failures if it fails to set file attributes like ownership
              '';
```

rnhmjoj commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -474,6 +474,11 @@ in
             };
+            options.securityModel = mkOption {
+              type = types.enum [ "passthrough" "mapped" "mapped-xattr" "mapped-file" "none" ];
+              default = "none";
```
It would be a good idea to change this to `mapped-xattr`. This is what qemu recommends and shouldn't cause issues, though it would require some testing.

ajamtli replied:
I agree that it would be sensible to use the recommended model as default. However with the proposed change it will change the security model for the nix-store as well. I'll look into it! 
Thanks!

rnhmjoj replied:
>  it will change the security model for the nix-store as well. I'll look into it!

Yeah, that's why we should be careful. I'll do some testing as well.

rnhmjoj commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -474,6 +474,11 @@ in
               description = lib.mdDoc "The mount point of the directory inside the virtual machine";
             };
+            options.securityModel = mkOption {
+              type = types.enum [ "passthrough" "mapped" "mapped-xattr" "mapped-file" "none" ];
```
```suggestion
              type = types.enum [ "passthrough" "mapped-xattr" "mapped-file" "none" ];
```
You can skip `mapped`, it seems to be an alias for backward compatibility.

ajamtli committed:
nixos/qemu-vm: added description of security model options, removed 'mapped' security model

Co-authored-by: Michele Guerini Rocco <rnhmjoj@users.noreply.github.com>

ajamtli committed:
nixos/qemu-vm: changed default security model

Changed the default security model for shared directories from 'none' to
the 'mapped-xattr'. QEMU recommends using this option for its security and
reliability benefits.

ajamtli committed:
nixos/qemu-vm: added security model for shared host nix store

Setting security model for shared host nix store to 'none'.
Using the default 'mapped-xattr' model results in a kernel panic when
running the VM.

eclairevoyant commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -474,6 +474,18 @@ in
+            options.securityModel = mkOption {
+              type = types.enum [ "passthrough" "mapped-xattr" "mapped-file" "none" ];
+              default = "mapped-xattr";
+              description = lib.mdDoc ''
```
```suggestion
              description = ''
```

ajamtli committed:
nixos/qemu-vm: removed use of lib.mdDoc

Co-authored-by: éclairevoyant <848000+eclairevoyant@users.noreply.github.com>

ajamtli committed:
Merge branch 'NixOS:master' into qemu-vm-virtfs-security-model

rnhmjoj commented:
Sorry for keeping you waiting. I just did some testing and the `mapped-xattr` mode is indeed causing some trouble.
Besides breaking the boot in the initrd (it looks like binaries in `/nix/store` don't have the executable bit), it's also causing all tests that use the `copy_from_vm` command (some permission issue with `/tmp/shared`).

I haven't figured out what's wrong, in fact the mapping using extended attributes appears to be working. Here's a proof:
![2024-04-20 13:18:35](https://github.com/NixOS/nixpkgs/assets/2817565/44a48c3d-357f-412b-ae18-2a9c3beee17d)

rnhmjoj commented:
After some digging I figured out the Nix build sandbox is [disabling](https://github.com/NixOS/nix/blob/8c4c2156bd786156a57219e78aa14a363ca8f041/src/libstore/unix/build/local-derivation-goal.cc#L1682) extended attributes as a security measure.
So, changing the default to `mapped-xattr` is ok as long as all directories that are part of the NixOS testing instrumentation are set explicitely to `none`. It shouldn't matter anyway, as those are short-lived files that only exist while the test is running.

ajamtli committed:
nixos/qemu-vm: set security model 'none' for shared xchg directory

Co-authored-by: Michele Guerini Rocco <rnhmjoj@users.noreply.github.com>

ajamtli commented:
> After some digging I figured out the Nix build sandbox is [disabling](https://github.com/NixOS/nix/blob/8c4c2156bd786156a57219e78aa14a363ca8f041/src/libstore/unix/build/local-derivation-goal.cc#L1682) extended attributes as a security measure. So, changing the default to `mapped-xattr` is ok as long as all directories that are part of the NixOS testing instrumentation are set explicitely to `none`. It shouldn't matter anyway, as those are short-lived files that only exist while the test is running.

Great work figuring that one out! Thanks a lot for your help!

rnhmjoj commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -1091,6 +1103,7 @@ in
+        securityModel = "none";
       };
       xchg = {
         source = ''"$TMPDIR"/xchg'';
```
Still missing two more `securityModel = "none";`....

ajamtli committed:
nixos/qemu-vm: set security model for 'xchg' directory to 'none'

Co-authored-by: Michele Guerini Rocco <rnhmjoj@users.noreply.github.com>

ajamtli committed:
Revert "nixos/qemu-vm: set security model for 'xchg' directory to 'none'"

This reverts commit c16ff7f9f3f15723f77d4db13ef6ad6bb7619933.

eclairevoyant commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -1091,9 +1103,12 @@ in
       xchg = {
         source = ''"$TMPDIR"/xchg'';
+        securityModel = "none";
+        securityModel = "none";
```
seems like it was added twice here, instead of to `virtualisation.sharedDirectories.shared` and `virtualisation.sharedDirectories.certs`

ajamtli committed:
nixos/qemu-vm: set secrurity model for 'shared' and 'certs' directories to 'none'

eclairevoyant commented:
@rnhmjoj I think the explicitly requested changes were addressed, but is there still some concern with the initrd?

rnhmjoj reviewed (approved):
Looks good. I'd run some more tests to be safe before merging, though.
